### ISSUE TITLE
Module strawman 3: Implicitly defined imports

### ISSUE THREAD

**Issue by munificent** [munificent/wren#122]
Now that #106 is implemented, Wren supports implicitly defined names at the top level. Given that, I think we can use that to simplify how modules could handle circular imports. This proposal would supercede #79, and fix #46.

The basic idea borrows heavily from Python. There, to resolve circular imports, you push some of the imports down later in the file. As long as you don't *use* the imported modules until after the imports execute, everything works.

Here's an example (as usual, syntax is just made up at this point):

```dart
// main.wren
export class Main {
  static inMain {
    IO.print("inMain")
  }
}

import Other
Other.inOther
```

First, we define a class `Main`. The `export` modifier says, "make this object the object exported by this module". In this case the object is the class object for the `Main` class.

Then we import some object from `Other`. This defines a top-level variable `Other` whose value is an object—an instance of some class—exported by the Other module.

Finally, we invoke a method on the object it imported from `Other`.

Here's the other module:

```dart
// other.wren
export class Other {
  static inOther {
    IO.print("inOther")
    Main.inMain
  }
}

import Main
```

This also exports a class. That defines a static method. In that method, we use `Main`, which hasn't been defined yet. Since it's a capitalized name, that's OK. It just gets implicitly declared.

After the class body, we execute the import and now `Main` has a real definition.

Executing this program starting from Main proceeds like so:

1. Start executing Main.
2. Define the `Main` class and store it as the export object for the module.
3. Get to the `import Other` statement. Since Other has not been loaded yet, pause executing Main and start loading and executing Other.
3. Execute the `Other` class definition and store it as the export object.
4. Execute the `import Main`. We've already started loading Main, so look it up and get its current export object. That's the `Main` class, so we bind `Main` to the Main class.
5. Other is done executing, so resume Main.
6. Get the object exported from Other (the `Other` class) and bind that to `Other` in Main.
6. Execute `Other.inOther`. We've got a reference to the `Other` class, so that's fine.
7. That in turn calls `Main.inMain`. Likewise, we've got that wired up now too, so it works as well.

Of course, in most cases, you won't have circular dependencies and you can have the imports up at the top where they naturally would go.

The basic pieces we need to support this are:

1. The VM keeps a map of module names to the previously-loaded code for that module. For each, we also store a reference to the object the module is exporting.

2. An `import` statement that loads a module and binds its exported object to a variable. (We don't technically need syntax for this, but it would probably be nice to have.) It needs to be able to pause executing one module while the other module is loaded.

3. A way to specify which object should be exported from a module. Again, syntax isn't strictly needed, but would probably help.

Unlike #79, this doesn't require any sort of monkey-patching of existing classes, and I think the resulting syntax and semantics are simpler.

Thoughts?

**Comment by zeckalpha** [comment]
Does this handle duplicate exports, i.e. the Other class is exported from two places?

**Comment by munificent** [comment]
Yes, it should. Think of it like the distinction between variables and values. A module exports an *object*, some value in memory. Another module imports that object and binds it to a variable with some name. There's also another "name", which is the identifier you use to specify which module you want to import. 

Conceptually, each of these is independent. You could have a module physically stored in `foo.wren` that exports a class named `Bar` that you import into a variable in your module named `Baz`. Of course, we'll have to decide to what degree the syntax lets you control each of these, but in principle they're decoupled.

So, it's entirely possible to have two modules that both export a class whose name happens to be `Other`. As long as those modules themselves have different ways of being identified/located, it shouldn't cause any problems.

**Comment by gsmaverick** [comment]
Any update on this and related bugs around importing code? Would love to help where I can to move this along as it'd be really awesome to finally have in Wren :)

**Comment by munificent** [comment]
No useful update yet. I started tinkering around with it some but I realized it's going to be more complex to implement than I originally thought.

Modules means instead of a single global scope, each module has its own top-level scope distinct from every other modules. That's a key feature, but it makes the implementation harder.

Right now, there are bytecode for loading and storing a global variable. The compiler generates those when it encounters usage of a top-level name. Interpreting those is pretty simple: we just look up the variables by index in the VM's global variable array.

When each module has its own globals, which module we look in depends on where the function being executed was defined. This naïvely means each function would have to have a pointer to its module. I really don't want to add another pointer to functions if I can avoid it since it impacts performance.

Fortunately, functions already have a convenient way to access variables defined outside of their scope: closures! We could just make functions close over the top-level variables they use. They would then end up in the upvalue array like any other outer variable. Swell.

That *may* have a negative performance impact, but at least it's only in the functions that do actually access top-level stuff. I'd need to experiment to see how bad it is.

But closing over them means a lot of changes in the compiler. What's particularly tricky is figuring out how it interacts with the REPL. Right now, the VM keeps around an array of the names of all global variables. This makes it much easier for the REPL to use variables defined in previous lines of code.

If those become more like regular scoped local variables, the compiler will "forget" about them after it's done compiling a line of code in the REPL, and being able to access them later gets trickier.

I don't think these problems are insurmountable at all, but it's going to be a bit of work, and I have to carve out a larger block of free time for that. Right now, it's easier for me to work on some smaller scale stuff. I totally agree this is hugely important, though.
